docs/rfcs/0003-sql-datastore.md

```diff
@@ -0,0 +1,62 @@
+- Start Date: 2020-02-22
+- Target Version: 1.0.0
+
+# Summary
+
+This RFC proposes adding SQL database as datastore for PipeCD Control-plane. Currently, we're considering between PostgreSQL and MySQL for those NoSQL support features.
+
+# Motivation
+
+The background of this proposal is we are going to add complex queries and more indexes to the control-plane datastore for some new features of PipeCD, the current in-use NoSQL datastore is blocking those requirements.
+
+# Detailed design
+
+As mention is the motivation section, the root cause of this change is: we're going to add complex queries with those indexes to the datastore of PipeCD control-plane. The use-case of those queries requires:
+- easy way to pagination query result
+- easy way to ordering query result
+- easy way to add indexes for future adding queries
+
+Besides, to keep `simplicity on installing` characteristic of PipeCD, the chosen must be:
+- easy to be installed on cloud provider environment (GCP, AWS, etc)
+- can use the updated version easily
+
+We're using NoSQL for now due to its schemaless characteristic is good for the deployment process, so we focus on NoSQL support features on the chosen SQL database.
+
+The sample table creates commands for the `project` model and `application` model (other models than project model is as same as the application) as below.
+
+```sql
+CREATE TABLE project (
+	id VARCHAR ( 32 ) PRIMARY KEY,
+	data JSON NOT NULL,
+	created_at TIMESTAMP NOT NULL,
+	updated_at TIMESTAMP NOT NULL,
+);
+
+CREATE TABLE application (
+	id VARCHAR ( 32 ) PRIMARY KEY,
+	project_id VARCHAR ( 32 ) NOT NULL,
+	data JSON NOT NULL,
+	created_at TIMESTAMP NOT NULL,
+	updated_at TIMESTAMP NOT NULL,
+);
+```
+
+For indexing issue:
+- PostgreSQL indexing for the attributes of Jsonb data type is as follow: [Json indexing](https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING)
+- MySQL also support indexing for attributes of Json data type as follow: [Json indexing by secondary indexes](https://dev.mysql.com/doc/refman/8.0/en/create-table-secondary-indexes.html)
+
+(note: both Jsonb and Json point to the type of storing JSON data in those databases. The key difference is that Json data is stored as an exact copy of the JSON input text, whereas Jsonb stores data in a decomposed binary form. Jsonb seems a better choice since it supports indexing with significant performance increases)
+
+# Alternatives
+
+Currently, we consider between MySQL and PostgreSQL for those support for NoSQL features, PostgreSQL has a longer time in this field.
```

Comment:
(Some thoughts about what should be chosen: MySQL or PostgresSQL)
Based on our data size and the way of using the database (just a simple feature) I think performance and feature-set are not big factors used to consider when choosing.
Better than those, the popularity would be a good factor. The popularity means which is more familiar to our users, supported well by many cloud providers, has a good library to use...

Btw, let's make this decision after having a clearer look at the implementation detail.
https://github.com/pipe-cd/pipe/pull/1612#issuecomment-784726947

Reply:
We majorly depend on the JSON field, so the ability of the JSON field supported by the database would be a good factor too. (The ability to index a specific attribute of the JSON field, the supported operator on that field, and the performance of operating on that field.)

Reply:
So interesting. I'm just curious about whether you guys are looking into supporting both in the end if both have enough abilities.